src/ast/node/function.rs

```diff
@@ -307,6 +297,98 @@ impl Node for FuncCallNode {
     }
 }
 
+pub fn generic_tp_apply<'a, 'b, T: Generic + CustomType, N: GenericInferenceAble + RangeTrait>(
+    t: &T,
+    n: &N,
+    ctx: &'b mut Ctx<'a>,
+    builder: &'b BuilderEnum<'a, '_>,
+) -> Result<(), PLDiag> {
+    *ctx.need_highlight.borrow_mut() += 1;
+    let generic_size = t.get_generic_size();
+    let generic_params = n.get_generic_params();
+    let range = n.range();
+    let mut generic_types = preprocess_generics(generic_params, generic_size, range, ctx, builder)?;
+
+    let re = ctx.run_in_type_mod(t, |ctx, t| {
+        ctx.protect_generic_context(t.get_generic_map(), |ctx| {
+            for (i, (_, pltype)) in t.get_generic_map().iter().enumerate() {
+                if i >= generic_size {
+                    break;
+                }
+                if generic_types[i].is_none() {
+                    // fill inferred types
+                    let ty = n
+                        .get_inference_result()
+                        .clone()
+                        .unwrap_or_default()
+                        .get(i)
+                        .map(|v| {
+                            let ty = ctx.unify_table.borrow_mut().probe(*v);
+                            ty.get_type(ctx, builder, &mut ctx.unify_table.clone().borrow_mut())
+                        });
+                    if let Some(ty) = ty {
+                        if *ty.borrow() != PLType::Unknown {
+                            generic_types[i] = Some(ty);
+                        } else {
+                            continue;
+                        }
+                    } else {
+                        continue;
+                    }
+                }
+                let res = ctx.eq(pltype.clone(), generic_types[i].as_ref().unwrap().clone());
+                if !res.eq {
+                    // let r = generic_params.generics[i].as_ref().unwrap().range();
+                    // return Err(r.new_err(ErrorCode::ILLEGAL_GENERIC_PARAM).add_to_ctx(ctx));
+                    let g = t.get_generic_map().get_index(i).unwrap();
+                    let mut diag = range.new_err(ErrorCode::ILLEGAL_GENERIC_PARAM);
+                    if let Some(reason) = res.reason {
+                        diag.add_help(&reason);
+                    }
+                    diag.add_label(
+                        g.1.borrow().get_range().unwrap_or_default(),
+                        t.get_path(),
+                        format_label!("parameter `{}` defined in `{}`", g.0, t.get_name()),
+                    );
+                    return Err(diag.add_to_ctx(ctx));
+                }
+            }
+            Ok(())
+        })
+    });
+    *ctx.need_highlight.borrow_mut() -= 1;
+    re
+}
+
+/// # preprocess_generics
+///
+/// This method will ensure that the number of generic parameters is correct.
+fn preprocess_generics<'a, 'b>(
+    generic_params: &Option<Box<GenericParamNode>>,
```

Comment:
the name of `generic_params` is misleading, as it iterates the field of `GeneicParamNode` and ensures the desired size.

Consider to rename it such as `param_node`